From bdafe351edde4e0005b5ea3c98f8c3c2a43cb08c Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 14 Nov 2022 11:38:15 +0800
Subject: [PATCH 0057/1398] [Huawei] Revert "memcg: Add static key for memcg
 kswapd"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187973

-------------------------------

This reverts commit 84355bcc1e8008181ed3e7275b2cd60a45d7476d.

A better solution of memcg async reclaim will be applied, just revert
origin memcg kswapd patchset.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/memcontrol.h |  3 ---
 mm/memcontrol.c            | 43 ++++++++------------------------------
 mm/vmscan.c                |  3 +--
 3 files changed, 10 insertions(+), 39 deletions(-)

diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index bddf3c28a2fe..bf29bd0ffda3 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -1903,9 +1903,6 @@ static inline void mem_cgroup_flush_foreign(struct bdi_writeback *wb)
 
 #endif	/* CONFIG_CGROUP_WRITEBACK */
 
-extern struct static_key_false memcg_kswapd_key;
-#define mem_cgroup_kswapd_enabled static_branch_unlikely(&memcg_kswapd_key)
-
 struct sock;
 bool mem_cgroup_charge_skmem(struct mem_cgroup *memcg, unsigned int nr_pages);
 void mem_cgroup_uncharge_skmem(struct mem_cgroup *memcg, unsigned int nr_pages);
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index a38b1196331d..5612ff2ffb68 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -97,10 +97,6 @@ bool cgroup_memory_noswap __read_mostly;
 static DECLARE_WAIT_QUEUE_HEAD(memcg_cgwb_frn_waitq);
 #endif
 
-static bool cgroup_memory_kswapd = false;
-DEFINE_STATIC_KEY_FALSE(memcg_kswapd_key);
-EXPORT_SYMBOL(memcg_kswapd_key);
-
 /* Whether legacy memory+swap accounting is active */
 static bool do_memsw_account(void)
 {
@@ -2372,15 +2368,10 @@ static void high_work_func(struct work_struct *work)
 {
 	struct mem_cgroup *memcg;
 
-	if (mem_cgroup_kswapd_enabled) {
-		current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
-		memcg = container_of(work, struct mem_cgroup, high_work);
-		reclaim_high(memcg, MEMCG_CHARGE_BATCH, GFP_KERNEL);
-		current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
-	} else {
-		memcg = container_of(work, struct mem_cgroup, high_work);
-		reclaim_high(memcg, MEMCG_CHARGE_BATCH, GFP_KERNEL);
-	}
+	current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
+	memcg = container_of(work, struct mem_cgroup, high_work);
+	reclaim_high(memcg, MEMCG_CHARGE_BATCH, GFP_KERNEL);
+	current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
 }
 
 /*
@@ -2550,17 +2541,11 @@ void mem_cgroup_handle_over_high(void)
 	 * memory.high is currently batched, whereas memory.max and the page
 	 * allocator run every time an allocation is made.
 	 */
-	if (mem_cgroup_kswapd_enabled) {
-		current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
-		nr_reclaimed = reclaim_high(memcg,
-					    in_retry ? SWAP_CLUSTER_MAX : nr_pages,
-					    GFP_KERNEL);
-		current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
-	} else {
-		nr_reclaimed = reclaim_high(memcg,
-					    in_retry ? SWAP_CLUSTER_MAX : nr_pages,
-					    GFP_KERNEL);
-	}
+	current->flags |= PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD;
+	nr_reclaimed = reclaim_high(memcg,
+				    in_retry ? SWAP_CLUSTER_MAX : nr_pages,
+				    GFP_KERNEL);
+	current->flags &= ~(PF_SWAPWRITE | PF_MEMALLOC | PF_KSWAPD);
 
 	/*
 	 * memory.high is breached and reclaim is unable to keep up. Throttle
@@ -5679,9 +5664,6 @@ mem_cgroup_css_alloc(struct cgroup_subsys_state *parent_css)
 	struct mem_cgroup *memcg, *old_memcg;
 	long error = -ENOMEM;
 
-	if (cgroup_memory_kswapd)
-		static_branch_enable(&memcg_kswapd_key);
-
 	old_memcg = set_active_memcg(parent);
 	memcg = mem_cgroup_alloc();
 	set_active_memcg(old_memcg);
@@ -7497,13 +7479,6 @@ static int __init cgroup_memory(char *s)
 }
 __setup("cgroup.memory=", cgroup_memory);
 
-static int __init memcg_kswapd(char *s)
-{
-	cgroup_memory_kswapd = true;
-	return 0;
-}
-__setup("memcg_kswapd", memcg_kswapd);
-
 /*
  * subsys_initcall() for memory controller.
  *
diff --git a/mm/vmscan.c b/mm/vmscan.c
index a397d19944c9..bba0abde8ca7 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2839,9 +2839,8 @@ static bool is_memcg_kswapd_stopped(struct scan_control *sc)
 	bool is_stop = false;
 	unsigned long stop_flag = 0;
 
-	if (!cgroup_reclaim(sc) || !mem_cgroup_kswapd_enabled)
+	if (!cgroup_reclaim(sc))
 		return false;
-
 	if (memcg->memory.max == PAGE_COUNTER_MAX)
 		stop_flag = memcg->memory.high / 6;
 	else
-- 
2.17.1

